canvas_sdk/data/data_access_layer_client.py

```diff
@@ -0,0 +1,23 @@
+"""
+Data Access Layer client.
+
+This module is primarily responsible for executing calls to the gRPC service so that such details
+are abstracted away from callers. The results of the functions in this module are protobufs which
+must be mapped to user-facing objects.
+"""
+
+import grpc
+
+from canvas_generated.data_access_layer.data_access_layer_pb2 import ID, Patient
+from canvas_generated.data_access_layer.data_access_layer_pb2_grpc import (
+    DataAccessLayerStub,
+)
+from settings import DAL_TARGET
+
+_CHANNEL = grpc.insecure_channel(DAL_TARGET)
```

Comment:
I'm not sure if channels actually need to be closed on shutdown or not. There's no complaint about it when the process ends like there is for the server (which required some special handling).

I've seen other examples like this from Googling from people that seem knowledgeable.

Reply:
I'm actually having seconds thoughts on this. I'm actually almost sure this won't work because it will block other async requests since they are all using this channel.

I don't like the alternative either, which is using a context manager to create a channel for every request.

This article recommends separate channels for "areas of high load": https://grpc.io/docs/guides/performance/#python

It also recommends using a channel pool, which I think is probably the best approach, but I haven't found any example of that in Python yet.

Reply:
I did some reading, and this actually looks like it is OK. My premise above about async blocking was incorrect as well. gRPC channels obtained from `insecure_channel` are threadsafe:

https://grpc.github.io/grpc/python/_modules/grpc.html#insecure_channel

And the performance guide I cited above recommends reusing stubs and channels whenever possible. So this setup does seem to align with the recommended approach.

Future enhancements could include adding more channels, for example if certain RPCs are used much more than others they could get their own dedicated channels.

@aduane Based on these findings, it seems that home-app should make use of channels and stubs in this manner as well, i.e. reusing channels and stubs rather than creating new ones for each gRPC request.

Reply:
How will we know when a change is justified? Delayed calls? Increased memory usage?

Reply:
I'm not entirely sure what it looks like when gRPC stalls due to channel initialization overhead. Maybe it blocks? It might require experimentation to know that for sure.